### PR TITLE
[FIX] web: utils/strings: sprintf support lazy translated strings

### PR DESCRIPTION
**Pull request by kebeclibre**
Have something like:
```js

sprintf("my string %s", _lt("lazy translated"));

```
Before this commit, the lazy translated term was not formatted into the string.
This was because _lt returns an object containing a custom toString function, which was not handled by sprintf

After this commit, sprintf is able to handle lazy translated strings

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/core/utils/strings.js [hunk 1]
```diff
@@ -71,12 +71,14 @@ export function intersperse(str, indices, separator = "") {
  */
 export function sprintf(s, ...values) {
     if (values.length === 1 && typeof values[0] === "object") {
-        const valuesDict = values[0];
-        s = s.replace(/\%\(?([^\)]+)\)s/g, (match, value) => valuesDict[value]);
-    } else if (values.length > 0) {
-        s = s.replace(/\%s/g, () => values.shift());
+        if (values[0].toString !== Object.prototype.toString) {
+            values = [`${values[0]}`];
```
Comment (aab-odoo) [inline addons/web/static/src/core/utils/strings.js:75]: add a comment here

**Comment by kebeclibre** [comment]
robodoo rebase-ff

File: addons/web/static/src/core/l10n/translation.js [hunk 2]
```diff
@@ -2,6 +2,31 @@
 
 export const translatedTerms = {};
 
+class LazyTranslatedString extends String {
+    valueOf() {
+        const str = super.valueOf();
+        return translatedTerms[str] || str;
```
Comment (aab-odoo) [inline addons/web/static/src/core/l10n/translation.js:8]: use _t()?

File: addons/web/static/tests/core/utils/strings_tests.js [hunk 3]
```diff
@@ -62,4 +64,37 @@ QUnit.module("utils", () => {
...
     });
+
+    QUnit.test("sprintf properly formats numbers", (assert) => {
+        assert.deepEqual(sprintf("Hello %s!", 5), "Hello 5!");
+        assert.deepEqual(sprintf("Hello %s and %s!", 9, 10), "Hello 9 and 10!");
+        assert.deepEqual(sprintf("Hello %(x)s!", { x: 11 }), "Hello 11!");
+        assert.deepEqual(sprintf("Hello %(x)s and %(y)s!", { x: 12, y: 13 }), "Hello 12 and 13!");
+    });
+
+    QUnit.test("sprintf set behavior when value is an Array", (assert) => {
```
Comment (aab-odoo) [inline addons/web/static/tests/core/utils/strings_tests.js:75]: sprintf set behavior?

File: addons/web/static/tests/core/utils/strings_tests.js [hunk 4]
```diff
@@ -62,4 +64,37 @@ QUnit.module("utils", () => {
...
+        assert.deepEqual(sprintf("Hello %s!", ["inarray"]), "Hello inarray!");
+        assert.deepEqual(sprintf("Hello %s and %s!", [9, "10"], [11]), "Hello 9,10 and 11!");
+        assert.deepEqual(sprintf("Hello %(x)s!", { x: [11] }), "Hello 11!");
+        assert.deepEqual(
+            sprintf("Hello %(x)s and %(y)s!", { x: [12], y: ["13"] }),
+            "Hello 12 and 13!"
+        );
+    });
```
Comment (aab-odoo) [inline addons/web/static/tests/core/utils/strings_tests.js:83]: weird to support that, but ok I guess. In the docstring we only talk about strings as values (or set of strings)

File: addons/web/static/src/core/utils/strings.js [hunk 5]
```diff
@@ -69,8 +69,9 @@ export function intersperse(str, indices, separator = "") {
  * @param {...string} ...values
  * @returns {string}
  */
+
```
Comment (aab-odoo) [inline addons/web/static/src/core/utils/strings.js:72]: sorry but could you remove this line?

**Comment by aab-odoo** [comment]
robodoo rebase-ff r+

**Pull request closed without merging**